### PR TITLE
Withdraw funds from bounty.

### PR DESCRIPTION
**Pull request by shamil-gadelshin**
### Changes
- added `withdraw_member_funding` extrinsic
- added `withdraw_creator_funding` extrinsic
- introduced bounty accounts.
- introduced bounty removal in some cases

### Implementation comments
- when the cherry can't be distributed fully among the funders - the remainder gets burned
- bounty state could be removed during either `withdraw_member_funding` or `withdraw_creator_funding`extrinsics. The actual removal occurs when all possible withdrawals are completed.

### Open questions
- Should we migrate a council budget to the account to eliminate slashes/deposit_creations? 
- How do we prevent bad bounty creation (eg.: funding_period = 0, creator_funding = 0)? Should we limit active bounties number for a member or we rely on the extrinsic fee? 

[Main issue](https://github.com/Joystream/joystream/issues/1998)

### CODE REVIEW

**Comment by bedeho** [comment]
> Should we migrate a council budget to the account to eliminate slashes/deposit_creations?

I did not understand this question, could you restate?

> How do we prevent bad bounty creation (eg.: funding_period = 0, creator_funding = 0)? Should we limit active bounties number for a member or we rely on the extrinsic fee?

If we have no caps on active bounties overall, then I see no need for such a limit.

File: runtime-modules/bounty/src/lib.rs [hunk 1]
```diff
@@ -194,19 +217,35 @@ pub enum BountyStage {
...
-    MaxFundingReached(BlockNumber),
+    /// Creator funds (initial funding and/or cherry) were withdrawn.
+    CreatorFundsWithdrawn,
+
+    /// A bounty funding was successful and it exceeded max funding amount.
+    BountyMaxFundingReached {
+        ///  A bounty funding was successful on the provided block.
+        max_funding_reached_at: BlockNumber,
+        /// A bounty reached its maximum allowed contribution on creation.
+        reached_on_creation: bool,
```
Comment (bedeho) [inline runtime-modules/bounty/src/lib.rs:239]: Why is `reached_on_creation` needed? I see it only read here

`cherry_needs_withdrawal: reached_on_creation && cherry_is_not_zero,`

but it's also not clear this field is needed, or why it depends on anything except `cherry_is_not_zero`.

File: runtime-modules/bounty/src/lib.rs [hunk 2]
```diff
@@ -194,19 +217,35 @@ pub enum BountyStage {
...
+    },
 
     /// A bounty was canceled.
     Canceled,
 
-    /// A bounty funding was successful on the provided block.
-    /// The stage is set when the funding exceeded max funding amount.
-    MaxFundingReached(BlockNumber),
+    /// Creator funds (initial funding and/or cherry) were withdrawn.
+    CreatorFundsWithdrawn,
```
Comment (bedeho) [inline runtime-modules/bounty/src/lib.rs:232]: Why is there a distinct extrinsic and thus milestone for withdrawing funding? All there is is bounty cancellation? 

See design doc
> The creator can cancel at any time before this, something that is of particular importance of in perpetual funding periods. If no one has contributed, any cherry is returned and the bounty is instantl removed, but if at least one contributor has funded, then the cherry will be split amont all contributors, and the bounty goes to the withdraw period.

So cancelling a bounty covers possible recovery of bounty contributed, if relevant. Right now you have two separate extrinsics and milestones. I am not actually sure you even need two `Canceled` and `CreatorFundsWithdrawn` even if we wanted two separate extrinsics, as the runtime should do the exact same thing going forward in both cases?

File: runtime-modules/bounty/src/benchmarking.rs [hunk 3]
```diff
@@ -172,66 +190,181 @@ benchmarks! {
...
+            ..Default::default()
+        };
+
+        Bounty::<T>::create_bounty(RawOrigin::Root.into(), params, Vec::new()).unwrap();
+
+        let bounty_id: T::BountyId = Bounty::<T>::bounty_count().into();
+
+        assert!(Bounties::<T>::contains_key(bounty_id));
+
+        Bounty::<T>::cancel_bounty(RawOrigin::Root.into(), creator.clone(), bounty_id).unwrap();
```
Comment (conectado) [inline runtime-modules/bounty/src/benchmarking.rs:322]: Can we add an assertion after this that the stage is `BountyStage::Withdrawal` with `cherry_needs_withdrawal == true`? (The same for  the other benchmark for `withdraw_creator_funding`.)

File: runtime-modules/bounty/src/benchmarking.rs [hunk 4]
```diff
@@ -172,66 +190,181 @@ benchmarks! {
...
     verify {
-        let bounty = <crate::Bounties<T>>::get(bounty_id);
+        let bounty = Bounty::<T>::bounties(bounty_id);
 
         assert!(matches!(bounty.milestone, BountyMilestone::Canceled));
         assert_last_event::<T>(Event::<T>::BountyVetoed(bounty_id).into());
     }
 
     fund_bounty {
+        let max_amount: BalanceOf<T> = 100.into();
```
Comment (conectado) [inline runtime-modules/bounty/src/benchmarking.rs:238]: Shouldn't we have a benchmark for each branch of `get_bounty_milestone_on_funding`?

As I understand currently we are only benchmarking the case `maximum_funding_reached` is true. Am I right?

**Review by conectado — CHANGES_REQUESTED** [review]
Overall benchmarking looks great, just left one minor comment and one important question.

File: runtime-modules/bounty/src/benchmarking.rs (hunk 3 above)
Comment (shamil-gadelshin) [reply runtime-modules/bounty/src/benchmarking.rs:322]: Ok

File: runtime-modules/bounty/src/benchmarking.rs (hunk 4 above)
Comment (shamil-gadelshin) [reply runtime-modules/bounty/src/benchmarking.rs:238]: This is the worst-case scenario because it fires an additional event. Other branches are similar.

**Review by conectado — APPROVED** [review]
LGTM

**Pull request merged**